rfcs/rfc-colin-states-mapping-to-lear.md

```diff
@@ -0,0 +1,154 @@
+
+# Detailed design
+
+Business GET endpoint will return currently actionable filings list.  This will be determined using the allowable filing transitions map.
+
+_Examples of what this might look like but will need to represent sub-types of filings where applicable_
+``` http request
+
+GET https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/CP0001199
+
+{
+    "business": {
+        "legalName": "XYZ TEST CORP.",
+        "legalType": "BC",
+        "state": "HISTORICAL",
+        "currentlyActionableFilings": {
+            "filingTypes": ["restoration", "putBackOn"],
+            "filingSubtypes": {
+               "restoration": ["fullRestoration", "limitedRestoration"]
+            }   
+        }
+        ...
+}
+```
+
+``` http request
+
+GET https://legal-api-dev.apps.silver.devops.gov.bc.ca/api/v2/businesses/CP0001199
+
+{
+    "business": {
+    "legalName": "XYZ TEST CORP.",
+    "legalType": "BC",
+    "state": "ACTIVE",
+    "currentlyActionableFilings": {
+        "filingTypes": ["annualReport", "changeOfDirector", "alteration", "dissolution"],
+        "filingSubtypes": {
+            "dissolution": ["voluntaryDissolution"]
+        }   
+    }
+    ...
+}
+```
+
+
+An invalid business state transition map will be implemented as a python dictionary or database table.  This map can be used in conjunction with the existing [ALLOWABLE_FILINGS](https://github.com/bcgov/lear/blob/e850f9a22672910db6e4ceb2f1ddb9437541a86f/legal-api/src/legal_api/services/authz.py#L96) dictionary as well as any other conditions to determine what filings are allowed for a business at a given point in time.
+
+_Example of what invalid state transition might look like_
+``` python
+INVALID_STATE_TRANSITION_FILINGS: Final = {
+    'staff': {
+        Business.State.ACTIVE: { # current business state
+            'restoration': { # target filing
+                # invalid target filing if filing that transitioned business to its current state is in list
+                'fullRestoration': ['registration', 'incorporationApplication'], 
+                'limitedRestoration': ['registration', 'incorporationApplication'], 
+            }
+            ...
+        },
+        Business.State.HISTORICAL: {
+            'restoration': { 
+                'fullRestoration': ['almagmation', 'continuationOut', 'continuationIn'], 
+                'limitedRestoration': ['almagmation', 'continuationOut', 'continuationIn'],
+            },
+            'putBackOn': ['almagmation', 'withdrawal']
+            ...
+        },
+        Business.State.LIQUIDATION: {
+            'restoration': {
+                'fullRestoration': ['liquidation'], 
+                'limitedRestoration': ['liquidation'], 
+            },
+            ...
+        }
+    },
+    'client': {
```

Comment:
we may have other classes of users, like Lawyers, that would be allowed to do actions that the general public cannot.
maybe this should be 'default' or 'general' or some other term?
some API users may also be given status to do other filings.
we are implementing _feature flags_ to block/allow certain users with the User. login_source == 'API_GW'
and yes, I do realize that our "CaSe" isn't consistent :(

Reply:
If the BE knows as much about the user/roles as the FE, then the BE should be able to provide a suitable list of "allowable actions" based on that, right? So you're saying that it's more than staff vs non-staff, but possibly another (or more) classes of users.

Reply:
Updated to `client` to `general`.